main/src/ca/uwaterloo/flix/language/phase/Instances.scala

```diff
@@ -27,19 +28,154 @@ import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps, Validation}
 object Instances extends Phase[TypedAst.Root, TypedAst.Root] {
 
   /**
-    * Validates instances in the given AST root.
+    * Validates instances and classes in the given AST root.
     */
   override def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = flix.phase("Instances") {
     for {
       _ <- visitInstances(root)
+      _ <- visitClasses(root)
     } yield root
   }
 
+  /**
+    * Validates all instances in the given AST root.
+    */
+  private def visitClasses(root: TypedAst.Root)(implicit flix: Flix): Validation[Unit, InstanceError] = {
+
+    /**
+      * Checks that every super class of `class0` is lawful, unless `class0` is marked `lawless`.
+      */
+    def checkLawfulSuperClasses(class0: TypedAst.Class): Validation[Unit, InstanceError] = class0 match {
+      case TypedAst.Class(_, mod, sym, _, superClasses, _, _, loc) =>
```

Comment:
As a philosophical point. Would it be better to either:
- Return the class inside the validation (so it does not return unit), or 
- Simply return a List of InstanceErrors (if the list is empty there are no errors).

I am also fine with the current code.

Reply:
> Return the class inside the validation (so it does not return unit)

I prefer not to return the class; returning unit makes it clear that no transformation is done, whereas if we return the class, we end up with two copies of the same value in the namespace:
```scala
class0 = ...
for {
  class1 <- checkLawfulSuperClasses(class0)
  // class1 same as class0
...
```
Granted, we can ignore the return value, but in general when I see a return value, my immediate thought is that I should be using it.

> Simply return a List of InstanceErrors (if the list is empty there are no errors).

I think this would make it harder to `flatMap` with other validations.